llvm/lib/Target/AIE/AIEBaseHardwareLoops.cpp

```diff
@@ -173,6 +173,54 @@ char AIEBaseHardwareLoops::ID = 0;
 INITIALIZE_PASS(AIEBaseHardwareLoops, DEBUG_TYPE, AIE_HARDWARE_LOOPS_NAME,
                 false, false)
 
+namespace {
+// Everything becomes much simpler if ZOLs exit to their layout successor
+// We can achieve that by splitting a non-fallthrough exit jump off into
+// a new fallthrough block.
+// Note that we don't actually check the block layout; we are near the end of
+// codegen pipeline, and we assume that explicit jump to the fallthrough don't
+// occur.
+bool splitLoopEndJump(MachineBasicBlock &MBB, const AIEBaseInstrInfo *TII) {
+  auto Terminator = MBB.getFirstInstrTerminator();
+  if (Terminator == MBB.end() ||
+      !TII->isHardwareLoopEnd((*Terminator).getOpcode())) {
+    return false;
+  }
+  SmallVector<MachineOperand, 4> Cond;
+  MachineBasicBlock *TBB, *FBB;
+  const bool AllowModify = false;
+  TII->analyzeBranch(MBB, TBB, FBB, Cond, AllowModify);
+  if (!FBB) {
```

Comment:
I don't remember how `analyzeBranch` works (who does?), but shouldn't we only trust `TBB` or `FBB` when the branch was successfully analyzed?

Reply:
Yes, that would be better. I will change this.